### ISSUE TITLE
Cannot make 2.1.1 work on NodeJS based TW5

### ISSUE THREAD

**Issue by mauloop** [joshuafontany/TW5-JsonMangler#12]
Environment: Windows 10 + NodeJS 12.13.0 + TW 5.1.22-pre

I tried to update JsonMangler plugin to the latest release, but NodeJS raises an error and crashes.

**Test 1:** 
added the following to the TW plugin folder

- ebalster/modloader 0.0.4 
- matabele/action-maketid 0.0.2
- joshuafontany/jsonmangler 2.1.1

The following error has been raised by NodeJS:

```
Error executing boot module $:/core/modules/wiki.js: {}

$:/core/modules/wiki.js:127
exports.dispatchEvent = function(type /*, args */) {
                                                 ^

SyntaxError: Unexpected token ')'
    at new Script (vm.js:84:7)
    at createScript (vm.js:258:10)
    at Object.runInNewContext (vm.js:299:10)
    at Object.$tw.utils.evalSandboxed (C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:518:5)
    at Object.$tw.modules.execute (C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:794:15)
    at C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:834:30
    at Object.$tw.utils.each (C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:135:12)
    at Object.$tw.modules.forEachModuleOfType (C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:833:12)
    at Object.$tw.modules.applyMethods (C:\Users\boero\MK\opt64\TiddlyServer-2.1.5\tiddlywiki\boot\boot.js:857:14)
    at Object.exports.startup ($:/core/modules/startup/load-modules.js:29:14)
```
**Test 2:**

I excluded joshuafontany/jsonmangler from tiddlywiki.info. The wiki has been loaded without problems.

**Test 3:** 

I added the plugin by dragging the link directly from your wiki to mine. It has been loaded (a page refresh was required), but the first time I had to restart NodeJS, it crashes with the same error as test 1

**Test 4:**
I noticed that in the zip package release you included the ebalster/attribute-modules plugin, despite it is not specified as a dependency in the documention. Anyway I added this one too. NodeJS crashed raising the followin error:

```
Error executing boot module $:/core/modules/utils/parseutils.js: {}

$:/core/modules/utils/parseutils.js:31
/*
^^

SyntaxError: Invalid or unexpected token
    at new Script (vm.js:84:7)
    at createScript (vm.js:258:10)
    at Object.runInNewContext (vm.js:299:10)
    at Object.$tw.utils.evalSandboxed (Z:\opt64\tw5.1.22-pre\boot\boot.js:530:5)
    at Object.$tw.modules.execute (Z:\opt64\tw5.1.22-pre\boot\boot.js:806:15)
    at Z:\opt64\tw5.1.22-pre\boot\boot.js:846:30
    at Object.$tw.utils.each (Z:\opt64\tw5.1.22-pre\boot\boot.js:135:12)
    at Object.$tw.modules.forEachModuleOfType (Z:\opt64\tw5.1.22-pre\boot\boot.js:845:12)
    at Object.$tw.modules.applyMethods (Z:\opt64\tw5.1.22-pre\boot\boot.js:869:14)
    at Object.exports.startup ($:/core/modules/startup/load-modules.js:21:14)
```

The only way I have to use plugin rel. 2.1.1 is to add it and its dependencies to a single page wiki not served by NodeJS, but I really need it to work in a NodeJS environment. I tested backward the previous releases and all of the have the same issue, since 2.0.8.

I'm currently running rel. 2.0.6 on TW 5.1.21. 

Am I doing something wrong? I would appreciate very much your help.

)+(

**Comment by joshuafontany** [comment]
Hi Mauloop,

Thanks for bringing this to my attention with such good notes. I will investigate.
I developed this plugin on Win10+Node.js with the Bob server plugin, so this is interesting. 

I am currently on Node v12.14.0. I'm going to refresh the installs and try to mirror your launch environment. 

From my own experience debugging, it seems that the "Patch" file that alters one of the core files has bugged, leading to a javascript parse-error (which can be hard to track down).

Are you using any other plugins that use Evan's patch tool? The Formula plugin? If that is on an old version (which are not compatible with 5.1.22-pre) that also might be doing it. I have updated Even's plugin within a repo on my own github here. Might be worth a shot.

Best,
Joshua F

**Comment by mauloop** [comment]
Hi, Joshua.

Thanks for the time you are spending with my request. My actual work environment is TiddlyServer 2.1.5, which in turns runs TW 5.1.21. This was the environment where I tried to update JsonMangler plugin at the beginning and where I first encountered the error.

My troubleshooting approach is to remove any unrelated component from my test configuration. So I reproduced the errors that I reported  running **a pure NodeJS TW server with a brand new wiki**, which has no tiddlers of any kind, and no other plugins, but the filesystem and tiddlyweb which are required for a bsic server edition.

For the test purpose I cloned 5.1.22-pre from the Github repo.

> Are you using any other plugins that use Evan's patch tool?

No, JsonMangler plugin is the only case.

> I have updated Even's plugin within a repo on my own github here. Might be worth a shot.

I was not able to find it. Can you please post the link here?

Best regards,

)+(

**Comment by joshuafontany** [comment]
Hi mauloop,

I think the issue here is that none of my Releases have patch files that target TW 5.1.21. (It goes from 5.1.20 to 5.1.22-pre.... Oops.)

Jeremy has just officially released TW 5.1.22, would you be able to upgrade and test that. I will be making some time this week to test TiddlyServer.

Best,
Joshua

**Comment by mauloop** [comment]
As I wrote I ran all of the test in an isolated environment with 5.1.22-pre. I read that 5.1.22 has been officially released (yuppie!). If you suggest to test against the final release I will do, but sincerely I cannot see what could have been changed?

)+(

**Comment by mauloop** [comment]
A little update. I have just installed TW 5.1.22 and run all of the test again.

**Environment:**
- OS: Windows 10
- NodeJS 12.13.0
- Tiddlywiki5 5.1.22
- test wiki was a brand new wiki with just  tiddlywiki.info and an empty tiddler folder
- tiddlywiki.info was based on server edition (no plugins, but tiddlyweb + filesystem + highlight)

**Plugins version:**

- ebalster/modloader 0.0.4
- matabele/action-maketid 0.0.2
- joshuafontany/jsonmangler 2.1.1

**Test 1:**
Added the following to tiddlywiki.info in the "plugin" section

```
		"ebalster/modloader",
		"matabele/action-maketid",
		"joshuafontany/jsonmangler"
```

_Result:_ NodeJS crashes while starting, the following error message has been printed to stdout:

```
Error executing boot module $:/core/modules/wiki.js: {}

$:/core/modules/wiki.js:127
exports.dispatchEvent = function(type /*, args */) {
                                                 ^

SyntaxError: Unexpected token ')'
    at new Script (vm.js:84:7)
    at createScript (vm.js:258:10)
    at Object.runInNewContext (vm.js:299:10)
    at Object.$tw.utils.evalSandboxed (Z:\opt64\tw5.1.22\boot\boot.js:530:5)
    at Object.$tw.modules.execute (Z:\opt64\tw5.1.22\boot\boot.js:806:15)
    at Z:\opt64\tw5.1.22\boot\boot.js:846:30
    at Object.$tw.utils.each (Z:\opt64\tw5.1.22\boot\boot.js:135:12)
    at Object.$tw.modules.forEachModuleOfType (Z:\opt64\tw5.1.22\boot\boot.js:845:12)
    at Object.$tw.modules.applyMethods (Z:\opt64\tw5.1.22\boot\boot.js:869:14)
    at Object.exports.startup ($:/core/modules/startup/load-modules.js:29:14)
```

**Test 2:**
Added the following to tiddlywiki.info in the "plugin" section (similar to "Test 1", but no JsonMangler plugin)

```
		"ebalster/modloader",
		"matabele/action-maketid",
```

_Result:_ NodeJS starts without errors and the wiki load succesfully

**Test 3:**
No additions to tiddlywiki.info. The three plugins have been installed dragging the links from https://joshuafontany.github.io/TW5-JsonMangler/#%24%3A%2Fplugins%2Fjoshuafontany%2Fjsonmangler%2Freadme%2FInstallation.

_Result:_ 
After the import has been confirmed and the browser page has been reloaded everything looked like if it was all right. I could open the Control Panel e verify that the plugins was loaded.

As soon as I restarted NodeJS, it crashed with the same error as in "Test 1".

**Test 4:**
Added the following to tiddlywiki.info in the "plugin" section (similar to "Test 1" + attribute-modules plugin)

```
		"ebalster/attribute-modules",
		"ebalster/modloader",
		"matabele/action-maketid",
		"joshuafontany/jsonmangler"
```

_Result:_ NodeJS crashes while starting, the following error message has been printed to stdout:

```
Error executing boot module $:/core/modules/utils/parseutils.js: {}

$:/core/modules/utils/parseutils.js:31
/*
^^

SyntaxError: Invalid or unexpected token
    at new Script (vm.js:84:7)
    at createScript (vm.js:258:10)
    at Object.runInNewContext (vm.js:299:10)
    at Object.$tw.utils.evalSandboxed (Z:\opt64\tw5.1.22\boot\boot.js:530:5)
    at Object.$tw.modules.execute (Z:\opt64\tw5.1.22\boot\boot.js:806:15)
    at Z:\opt64\tw5.1.22\boot\boot.js:846:30
    at Object.$tw.utils.each (Z:\opt64\tw5.1.22\boot\boot.js:135:12)
    at Object.$tw.modules.forEachModuleOfType (Z:\opt64\tw5.1.22\boot\boot.js:845:12)
    at Object.$tw.modules.applyMethods (Z:\opt64\tw5.1.22\boot\boot.js:869:14)
    at Object.exports.startup ($:/core/modules/startup/load-modules.js:21:14)
```

I intended that you have a modified version of "ebalster/modloader". Is it correct? Do you think that could make sense to run the tests against this one?

)+(

P.S.: just to save your time...  the results are the same with TiddlyServer as with a vanilla NodeJS TW server. IMO, it makes no sense to spend time investigating TiddlyServer

**Comment by joshuafontany** [comment]
Hi mauloop,

Thank so much for the thorough testing, this has revealed a bug in my Dev environment.
Its a little esoteric, and involved how I was storing the "patch" for the main wiki.js module.

I have recreated your bug with both vanilla Node.js and Bob (finally). Bob was actually using my "patchSource" (the marked up copy of vanilla 5.1.20) AS a new wiki-module, overwritting the base 5.1.22. This mean, in certain situations like when loaded through Bob handling the booting, my test wiki worked (and any saved single-file versions worked, and reported v5.1.22-pre when actually all the internals were from 5.1.20).

WHOA, complex bug. Nice. Kudos. I noticed that the base wiki.js module in TW5.1.22 has actually changed a lot since I last generated a patch file.

I have corrected the meta-data in my "patchSource" file to prevent it being loaded a a WikiModule. I will need to manually markup the new 5.122 wiki.js module with my changes to link in the JsonMangler utils at the right spots. Then generate a new version of the plugin with a new "patch" file.

I have also used this opportunity to strip out some of the optional tools into their own plugin. This will make turn-around and testing a bit faster. Please hold for incoming build.

:)

Best,
Joshua F

**Comment by joshuafontany** [comment]
Hi @mauloop

Please try the new Release v2.2.0 (Official TW5.1.22 release):

https://github.com/joshuafontany/TW5-JsonMangler/releases/tag/v2.2.0

(or from the demo wiki). I have tested with both Bob and Node.js and it seems fixed.

Looking forward to your results.

**Comment by mauloop** [comment]
Dear Joshua,

I'm sorry to say that testing against rel. 2.2.0 returns the same error as in "Test 1" from my previous update.

As a workaround:
- from [your demo site](https://joshuafontany.github.io/TW5-JsonMangler/), exported the modded version of `$:/core/modules/wiki.js` as `$__core_modules_wiki.js.tid` 
- copied `$__core_modules_wiki.js.tid` in the tiddlers folder of my wiki
- copied `plugin.info` for rel. 2.2.0 of JsonMangler plugin to `$TW5/plugins/joshuafontany/jsonmangler`
- added just `joshuafontany/jsonmangler` to plugins section in `tiddlywiki.info` (no more `ebalster/modloader`)

Basically I override the core wiki.js module and get rid of modloader plugin. NodeJS starts without errors and the wiki loads successfully. What’s not there won’t break ;^)))

I'm running some tests with the JsonMangler plugin and I will update you.

)+(

**Comment by joshuafontany** [comment]
Thanks!

I hadn't mentioned it (but did update the readme), but I have removed the
dependency of 'maketid' and it is no longer bundled with the releases (it
was only necessary for the Control Panel add-on, and that's going to become
its own plugin).

Modloader & Jsonmangler would be the minimum test. Really neat Work Around!

**Comment by mauloop** [comment]
Hi, Joshua.

Just a little update about my workaround.

Instead of copying the modded wiki.js in the tiddlers folder of my wiki, I put it besides `plugin.info` in `$TW5/plugins/joshuafontany/jsonmangler`. This way `$:/core/modules/wiki.js` appears like a shadow tiddler contained in the JsonMangler plugin.

This is an hack for the plugin itself, without actually repackaging it. It works on vanilla NodeIS server as well on TiddlyServer, giving me a chance to apply the hack once and make it available to all of the TW5 instances served by TS, without the need to add the modded wiki.js to every wiki.

Regards,

)+(

**Comment by mauloop** [comment]
One more update. Just imported all of the test tiddlers from your demo wiki to mine. Every test has been passed succesfully (except, of course, the ones related to plugin packing feature, since it has been removed). Also successfully tested a 7000+ rows x 20 columns CSV.

It looks like if the hack is really working!

The only bug I noticed, which also affects your online wiki, is an annoying warning popup every time I manually edit a tiddler of type "application/json". This did not happen with rel 2.0.6 on TW5.1.21.

Regards,

)+(

**Comment by joshuafontany** [comment]
I believe I have tracked this down. The wiki.js "patchSource" file was encoded with Windows line-feed (CRLF), and my local Dev TW repo also had Windows encoding, but when compared/patched in Linux line-feed (LF) environment it would fail.

I have normalized all my repos that use this patching method to LR, and regenerated all the affected files (the patches in JsonMangler and Attribute-Modules). A new release, v2.2.1, has been posted.

**Comment by mauloop** [comment]
Hi, Joshua.

I'm running TW5 on Windows too, so it should not be affected by the line termination mismatch. Anyway I will test v2.2.1 as soon as possible. I am a little busy in these days.

Thanks for the update.

)+(
